Support `CASE` function with Calcite

LantaoJin opened the pull request:
### Description
[Describe what this change achieves]

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
<!-- List any other related issues here -->

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
 - [ ] New functionality has javadoc added.
 - [ ] New functionality has a user manual doc added.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

LantaoJin commented on ppl/src/main/java/org/opensearch/sql/ppl/parser/AstExpressionBuilder.java:
```diff
@@ -237,6 +238,29 @@ public UnresolvedExpression visitBooleanFunctionCall(BooleanFunctionCallContext
+                })
+            .collect(Collectors.toList());
+    UnresolvedExpression elseValue = new Literal(null, DataType.NULL);
+    if (ctx.caseFunction().valueExpression().size()
+        > ctx.caseFunction().logicalExpression().size()) {
+      // else value is present
+      elseValue =
+          visit(
+              ctx.caseFunction().valueExpression(ctx.caseFunction().valueExpression().size() - 1));
+    }
```
sure, I just copy from spark-ppl. will fix it. And I found whatever in ppl or sql. the `else` clause should be optional. If there is no `else` clause and no conditions are true, it returns NULL.

LantaoJin replied:
done

penghuo commented:
Update doc? https://github.com/opensearch-project/sql/blob/main/docs/user/ppl/functions/condition.rst

LantaoJin commented:
> Update doc? https://github.com/opensearch-project/sql/blob/main/docs/user/ppl/functions/condition.rst

Do you know the reason to make documentation executable? It's hard to debug if doctest fails. @penghuo

penghuo commented:
> Do you know the reason to make documentation executable? It's hard to debug if doctest fails. @penghuo

doctest has grown to have three primary uses, more reading https://github.com/opendistro-for-elasticsearch/sql/pull/497
* Checking examples in docstrings.
* Regression testing.
* Executable documentation / literate testing.

penghuo commented on docs/user/ppl/functions/condition.rst:
```diff
@@ -192,3 +192,49 @@ Example::
+    os> source=accounts | eval result = case(age > 35, firstname, age < 30, lastname else employer) | fields result, firstname, lastname
+    fetched rows / total rows = 4/4
+    +--------+-----------+----------+
+    | result | firstname | lastname |
```
add age and employer, `fields result, age, firstname, lastname, employer`

LantaoJin replied:
done

penghuo commented on docs/user/ppl/functions/condition.rst:
```diff
@@ -192,3 +192,49 @@ Example::
+    | null   | Dale      | Adams    |
+    +--------+-----------+----------+
+
+    os> source=accounts | eval result = case(age > 35, firstname, age < 30, lastname) | fields result, firstname, lastname
```
ditto, add age

LantaoJin replied:
done

LantaoJin commented:
I will backport this to branch 3.0 since it should be treated as bug:
1. CASE function is a supported function in SQL, this PR not only fixed it with v3, but also fixed it in v3.
2. This PR also fixed the bug of string literal which is critical for v3.

Can you approve it? @penghuo @anasalkouz @model-collapse
Also add related requirement link: https://github.com/opensearch-project/sql/issues/3523#issuecomment-2797219593

model-collapse commented:
Approved to backport
